File: src/rl_agent_python/full_test.py
"""
Full backtesting script for reinforcement learning trading strategy.

This script implements a rolling window backtesting approach to evaluate
the performance of reinforcement learning agents on cointegrated pairs.
"""

from dataclasses import dataclass
from typing import List, Tuple

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

from src.rl_agent_python.attributes import attributes_all
from src.rl_agent_python.cointegration import (
    cointegration_spreads,
    get_cointegrated_pairs,
)
from src.rl_agent_python.financial_functions import calculate_returns, ret_asset
from src.rl_agent_python.mcc import mcc_bootstrap
from src.rl_agent_python.pre_run import pre_run


@dataclass
class BacktestResult:
    """Class to store backtesting results."""

    agent_returns: np.ndarray
    pair_returns: np.ndarray
    train_periods: List[Tuple[int, int]]
    test_periods: List[Tuple[int, int]]


def plot_returns(agent_returns: np.ndarray, pair_returns: np.ndarray) -> None:
    """
    Plot agent and pair returns.

    Args:
        agent_returns: Array of agent returns
        pair_returns: Array of pair returns
    """
    plt.figure(figsize=(12, 6))
    plt.plot(agent_returns, label="Agent Returns")
    plt.plot(pair_returns, label="Pair Returns")
    plt.title("Rolling Window Backtest Results")
    plt.xlabel("Time")
    plt.ylabel("Returns")
    plt.legend()
    plt.grid(True)
    plt.show()


def rolling_window_backtest(
    df_currencies: pd.DataFrame,
    currencies: List[str],
    train_length: int = 3000,
    test_length: int = 200,
    n_episodes: int = 2,
    cost: float = 0,
    delay: int = 15,
) -> BacktestResult:
    """
    Perform rolling window backtesting.

    Args:
        df_currencies: DataFrame of currency data
        currencies: List of currency codes
        train_length: Length of training window
        test_length: Length of testing window
        n_episodes: Number of episodes for training
        cost: Transaction cost
        delay: Delay for attribute calculation

    Returns:
        BacktestResult object containing returns and periods
    """
    all_agent_returns = []
    all_pair_returns = []
    train_periods = []
    test_periods = []

    previous_agent_ret = 1.0
    previous_pair_ret = 1.0

    i = 1
    while (train_length + test_length * i) < len(df_currencies):
        # Define periods for this iteration
        train_start = test_length * (i - 1) + 1
        train_end = train_length + test_length * (i - 1)
        test_start = train_length + test_length * (i - 1) + 1
        test_end = train_length + test_length * i

        train_period = (train_start, train_end)
        test_period = (test_start, test_end)

        train_periods.append(train_period)
        test_periods.append(test_period)

        # Get cointegrated pairs
        coint_pairs = get_cointegrated_pairs(
            df_currencies, currencies, train_period, critical_value=3
        )

        # Calculate spreads
        coint_spreads = cointegration_spreads(
            coint_pairs, df_currencies, train_period, test_period
        )

        train_spreads = pd.DataFrame(coint_spreads[0])
        test_spreads = pd.DataFrame(coint_spreads[1])
        all_spreads = pd.DataFrame(coint_spreads[2])
        coefs = pd.DataFrame(coint_spreads[3])

        # Create attributes
        attributes_dict = attributes_all(all_spreads, all_spreads.columns, delay)

        # Backtest each cointegrated pair
        agent_returns = {}
        pair_returns = {}

        for j in range(len(coint_pairs.columns)):
            print(f"Processing pair {j+1}/{len(coint_pairs.columns)}")

            pair_idx = j
            coint_pair = coint_pairs.iloc[:, pair_idx]
            coef = coefs.iloc[:, pair_idx]

            # Prepare training data
            train_pair = df_currencies.iloc[
                train_period[0] - 1 : train_period[1], coint_pair
            ]
            train_spread = train_spreads.iloc[:, pair_idx]
            train_features = pd.DataFrame(attributes_dict[pair_idx]).iloc[
                train_period[0] - 1 : train_period[1]
            ]

            # Prepare testing data
            test_pair = df_currencies.iloc[
                test_period[0] - 1 : test_period[1], coint_pair
            ]
            test_spread = test_spreads.iloc[:, pair_idx]
            test_features = pd.DataFrame(attributes_dict[pair_idx]).iloc[
                test_period[0] - 1 : test_period[1]
            ]

            # Train and test MCC agent
            pretrained_agent = mcc_bootstrap(
                train_features, train_pair, coef, cost, n_episodes, algorithm="mcc"
            )

            pretrained_pos = pretrained_agent.Position_track
            pretrain_return = calculate_returns(
                train_pair, coef, pretrained_pos[-len(train_pair) :], cost
            )

            tested_agent = mcc_bootstrap(
                test_features,
                test_pair,
                coef,
                cost,
                n_episodes,
                pretrained_agent=pretrained_agent,
            )

            test_pos = tested_agent.Position_track
            test_return = calculate_returns(
                test_pair,
                coef,
                test_pos[-len(test_pair) :],
                cost,
                previous_ret=previous_agent_ret,
            )

            pair_name = f"{coint_pair.iloc[0]}_{coint_pair.iloc[1]}"
            agent_returns[pair_name] = test_return

            # Calculate pair returns
            pair_ret = pd.DataFrame(
                {
                    col: ret_asset(test_pair[col], previous_ret=previous_pair_ret)
                    for col in test_pair.columns
                }
            ).mean(axis=1)

            pair_returns[pair_name] = pair_ret

        # Calculate mean returns for this window
        agent_returns_df = pd.DataFrame(agent_returns)
        pair_returns_df = pd.DataFrame(pair_returns)

        window_agent_returns = agent_returns_df.mean(axis=1)
        window_pair_returns = pair_returns_df.mean(axis=1)

        all_agent_returns.extend(window_agent_returns)
        all_pair_returns.extend(window_pair_returns)

        previous_agent_ret = window_agent_returns.iloc[-1]
        previous_pair_ret = window_pair_returns.iloc[-1]

        i += 1

    return BacktestResult(
        agent_returns=np.array(all_agent_returns),
        pair_returns=np.array(all_pair_returns),
        train_periods=train_periods,
        test_periods=test_periods,
    )


def main():
    # Initialize data
    currencies = [
        "EURGBP",
        "EURUSD",
        "USDJPY",
        "EURJPY",
        "CHFJPY",
        "GBPUSD",
        "AUDJPY",
        "EURCAD",
    ]

    # Get currency data
    df_currencies = pre_run(currencies)

    # Run backtesting
    results = rolling_window_backtest(
        df_currencies,
        currencies,
        train_length=3000,
        test_length=200,
        n_episodes=2,
        cost=0,
        delay=15,
    )

    # Plot results
    plot_returns(results.agent_returns, results.pair_returns)


if __name__ == "__main__":
    main()

File: code/R/src/rl_agent/full_test.R
# A Reinforcement Learning Application for Statistical Arbitrage
# ATRGStrategies

# -----------------------------------------------------------------
# import dependencies (files and libraries)

#install.packages("plyr")
library(plyr) # data manipulations
setwd("C:/Users/user/Desktop/ATRG/ROO/atrg-strategies/src/main/R")
data_path <- paste(getwd(), c("/data/1min/"), sep ="")

source(file = "getdata.R")
source(file = "cointegration.R")
source(file = "attributes.R")
source(file = "financial_functions.R")
source(file = "discretization.R")
source(file = "state_space.R")
source(file = "reward.R")
source(file = "MCC.R")

# -----------------------------------------------------------------
#region Raw data


#c_currencies <- list.files("C:/Users/user/Desktop/UCL final project/data/1min") # 1min data
c_currencies <- c("EURGBP", "EURUSD", "USDJPY", "EURJPY", "CHFJPY", "GBPUSD", "AUDJPY", "EURCAD")
#c_currencies <- randomObs(c_currencies, 12) # 12 randon currencies
#c_currencies <- c_currencies[! c_currencies %in% c(c_currencies1)]
df_currencies <- GetQuotes(c_currencies, data_path, ".csv")


#endregion
# -----------------------------------------------------------------
#region Cointegration extraction

train_length <- 3000
test_length <- 200

Nepisodes <- 2
cost <- 0

i_delay <- 15

all_agent_mean_ret <- c()
all_asset_mean_ret <- c()
previous_agent_ret <- 1
previous_pair_ret <- 1

i <- 1
while((train_length + test_length*i) < nrow(df_currencies)){

  # After each test_length data points reestimate cointegration
  # on a window of length train_length
  train_period <- c(test_length*(i-1) + 1, train_length + test_length*(i-1))
  test_period <- c(train_length + test_length*(i-1) + 1, train_length + test_length*i)

  coint_pairs <- GetCointPairs(df_currencies, c_currencies, train_period, 3)
  coint_spreads <- CointSpreads(coint_pairs, df_currencies, train_period, test_period)

  train_spreads <- as.data.frame(coint_spreads[[1]])
  test_spreads <- as.data.frame(coint_spreads[[2]])
  all_spreads <- as.data.frame(coint_spreads[[3]])
  coefs <- as.data.frame(coint_spreads[[4]])

  l_attributes <- AttributesAll(all_spreads, colnames(all_spreads), i_delay)

  # Backtest each cointegrated pair
  agent_return <- list()
  pair_return <- list()

  for(j in 1:ncol(coint_pairs)){

    print(j / ncol(coint_pairs))

    pair_ind <- j
    coint_pair <- coint_pairs[,pair_ind]
    coef <- coefs[,pair_ind]

    train_pair <- df_currencies[c(train_period[1]:train_period[2]), coint_pair]
    train_spread <- train_spreads[[pair_ind]]
    train_features <- as.data.frame(l_attributes[pair_ind])[c(train_period[1]:train_period[2]),]

    test_pair <- df_currencies[c(test_period[1]:test_period[2]), coint_pair]
    test_spread <- test_spreads[[pair_ind]]
    test_features <- as.data.frame(l_attributes[pair_ind])[c(test_period[1]:test_period[2]),]

    #endregion
    # -----------------------------------------------------------------
    #region MCC test

    pretrained_agent <- MCCbootstrap(train_spread, train_pair, coef, train_features, Nepisodes, cost)
    pretrained_pos <- pretrained_agent[[2]]
    pretrain_return <- Ret(train_pair, coef, train_spread, tail(pretrained_pos, nrow(train_pair)), 0)
    #PlotAgentResult(train_pair, train_spread, pretrain_return)

    tested_agent <- MCCbootstrap(test_spread, test_pair, coef, test_features, Nepisodes, cost, pretrained_agent)
    test_pos <- tested_agent[[2]]
    test_return <- Ret(test_pair, coef, test_spread, tail(test_pos, nrow(test_pair)), 0, previous_ret=previous_agent_ret)
    agent_return[[paste(coint_pairs[1,j],coint_pairs[2,j])]] <- test_return

    #O jeden za duzo rowMeans w tym miesjcu?
    pair_return[[paste(coint_pairs[1,j],coint_pairs[2,j])]] <- rowMeans(apply(test_pair, 2, ret_asset,
                                                                              previous_ret=previous_pair_ret))
  }
  agent_return <- rowMeans(as.data.frame(agent_return))
  previous_agent_ret <- tail(agent_return, 1)
  all_agent_mean_ret <- c(all_agent_mean_ret, agent_return)


  pair_return <- rowMeans(as.data.frame(pair_return))
  previous_pair_ret <- tail(pair_return, 1)
  all_asset_mean_ret <- c(all_asset_mean_ret, pair_return)

  i <- i + 1
}


plot.ts(all_asset_mean_ret)
plot.ts(all_agent_mean_ret)
